core/base/v7/inc/ROOT/RError.hxx

```diff
@@ -0,0 +1,202 @@
+/// \file ROOT/RError.h
+/// \ingroup Base ROOT7
+/// \author Jakob Blomer <jblomer@cern.ch>
+/// \date 2019-12-11
+/// \warning This is part of the ROOT 7 prototype! It will change without notice. It might trigger earthquakes. Feedback
+/// is welcome!
+
+/*************************************************************************
+ * Copyright (C) 1995-2015, Rene Brun and Fons Rademakers.               *
+ * All rights reserved.                                                  *
+ *                                                                       *
+ * For the licensing terms see $ROOTSYS/LICENSE.                         *
+ * For the list of contributors see $ROOTSYS/README/CREDITS.             *
+ *************************************************************************/
+
+#ifndef ROOT7_RError
+#define ROOT7_RError
+
+#include <ROOT/RConfig.hxx>
+
+#include <cerrno>
+#include <cstddef>
+#include <new>
+#include <stdexcept>
+#include <string>
+
+namespace ROOT {
+namespace Experimental {
+
+// clang-format off
+/**
+\class ROOT::Experimental::RException
+\ingroup Base
+\brief Base class for all ROOT issued exceptions
+*/
+// clang-format on
+class RException : public std::runtime_error {
+public:
+   explicit RException(const std::string &what)
+      : std::runtime_error(what + " [errno: " + std::to_string(errno) + "]") {}
+};
+
+
+namespace Detail {
+
+// clang-format off
+/**
+\class ROOT::Experimental::Detail::RStatusType
+\ingroup Base
+\brief Values of this type have a distinct value that indicates an error, such as -1 for system calls
```

Comment:
You may want to call out to the fact that for the 'default' implementation 'all' values are consider an error.

Reply:
Changed to curiously recurring template pattern.